tests/unit/data/test_vector_factory.py

```diff
@@ -0,0 +1,115 @@
+        actual = VectorFactory.build(d)
+        expected = Vector(id='1', values=[0.1, 0.2, 0.3], metadata={'genre': 'comedy'}, sparse_values=SparseValues(indices=[0, 2], values=[0.1, 0.3]))
+        assert actual == expected
+
+    def test_build_when_dict_sparse_values_errors_when_not_dict(self):
+        with pytest.raises(ValueError, match="Column `sparse_values` is expected to be a dictionary"):
+            d = {'id': '1', 'values': [0.1, 0.2, 0.3], 'metadata': {'genre': 'comedy'}, 'sparse_values': 'not a dict'}
+            VectorFactory.build(d)
+
+    def test_build_when_dict_sparse_values_errors_when_missing_indices(self):
+        with pytest.raises(ValueError, match="Missing required keys in data in column `sparse_values`"):
+            d = {'id': '1', 'values': [0.1, 0.2, 0.3], 'metadata': {'genre': 'comedy'}, 'sparse_values': {'values': [0.1, 0.3]}}
+            VectorFactory.build(d)
+
+    def test_build_when_dict_sparse_values_errors_when_missing_values(self):
+        with pytest.raises(ValueError, match="Missing required keys in data in column `sparse_values`"):
+            d = {'id': '1', 'values': [0.1, 0.2, 0.3], 'metadata': {'genre': 'comedy'}, 'sparse_values': {'indices': [0, 2]}}
+            VectorFactory.build(d)
+
+    def test_build_when_dict_sparse_values_errors_when_indices_not_list(self):
+        with pytest.raises(ValueError, match="Found unexpected data in column `sparse_values`"):
+            d = {'id': '1', 'values': [0.1, 0.2, 0.3], 'metadata': {'genre': 'comedy'}, 'sparse_values': {'indices': 'not a list', 'values': [0.1, 0.3]}}
+            VectorFactory.build(d)
+
+    def test_build_when_dict_sparse_values_errors_when_values_not_list(self):
+        with pytest.raises(ValueError, match="Found unexpected data in column `sparse_values`"):
+            d = {'id': '1', 'values': [0.1, 0.2, 0.3], 'metadata': {'genre': 'comedy'}, 'sparse_values': {'indices': [0, 2], 'values': 'not a list'}}
+            VectorFactory.build(d)
+
+    def test_build_when_dict_sparse_values_when_values_is_ndarray(self):
+        d = {'id': '1', 'values': [0.1, 0.2, 0.3], 'metadata': {'genre': 'comedy'}, 'sparse_values': {'indices': [0, 2], 'values': np.array([0.1, 0.3])}}
+        actual = VectorFactory.build(d)
+        expected = Vector(id='1', values=[0.1, 0.2, 0.3], metadata={'genre': 'comedy'}, sparse_values=SparseValues(indices=[0, 2], values=[0.1, 0.3]))
+        assert actual == expected
+
+    def test_build_when_dict_sparse_values_when_indices_is_ndarray(self):
+        d = {'id': '1', 'values': [0.1, 0.2, 0.3], 'metadata': {'genre': 'comedy'}, 'sparse_values': {'indices': np.array([0, 2]), 'values': [0.1, 0.3]}}
+        actual = VectorFactory.build(d)
+        expected = Vector(id='1', values=[0.1, 0.2, 0.3], metadata={'genre': 'comedy'}, sparse_values=SparseValues(indices=[0, 2], values=[0.1, 0.3]))
+        assert actual == expected
```

Comment:
Can users also pass a pandas array for `indices` / `values` and should that be tested as well?

Reply:
Yeah, they can since `convert_to_list` handles both. I'll add a few more tests.